consulnameresolver/src/main/java/io/datanerds/grpc/ConsulNameResolver.java

```diff
@@ -0,0 +1,122 @@
+package io.datanerds.grpc;
+
+import com.google.common.base.Preconditions;
+import com.google.common.net.HostAndPort;
+import com.orbitz.consul.Consul;
+import com.orbitz.consul.HealthClient;
+import com.orbitz.consul.cache.ServiceHealthCache;
+import com.orbitz.consul.cache.ServiceHealthKey;
+import com.orbitz.consul.model.ConsulResponse;
+import com.orbitz.consul.model.health.ServiceHealth;
+import com.orbitz.consul.option.ImmutableCatalogOptions;
+import io.grpc.*;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.InetSocketAddress;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+public class ConsulNameResolver extends NameResolver {
+    public static final String DEFAULT_ADDRESS = "localhost";
+    public static final Integer DEFAULT_PORT = 8383;
+    private static final Logger logger = LoggerFactory.getLogger(ConsulNameResolver.class);
+    private final String authority;
+    private final String service;
+    private Listener listener;
+    private Consul consul;
+    private HealthClient healthClient;
+    private Optional<String> datacenter;
+    private Optional<List<String>> tags;
+
+    public ConsulNameResolver(ConsulQueryParameter parameter) {
+        this.authority = parameter.consulAddress.orElse(
+                String.format("%s:%s", DEFAULT_ADDRESS, DEFAULT_PORT)); //todo move to URI
+        this.service = parameter.service;
+        this.datacenter = parameter.datacenter;
+        this.tags = parameter.tags;
+
+        HostAndPort consulHostAndPort = HostAndPort.fromString(this.authority);
+        this.consul = Consul.builder().withHostAndPort(consulHostAndPort).build();
+        this.healthClient = this.consul.healthClient();
+    }
+
+    @Override
+    public String getServiceAuthority() {
+        return this.authority;
+    }
+
+    @Override
+    public void shutdown() {
+    }
+
+    @Override
+    public void start(Listener listener) {
+        Preconditions.checkState(this.listener == null, "already started");
+        this.listener = Preconditions.checkNotNull(listener, "listener");
```

Comment:
error message "listener"? Should probably be more informative: "listener cannot be empty".